opentracing/src/main/java/opentracing/TraceContext.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+/**
+ * Represents the identity of a given Span and Trace Attributes. The span identity is embedded in
+ * this context in implementation-specific format. The contents are serialized on the wire during
+ * inter-process calls. Trace Attributes can be added to the context and will be available to all
+ * future children of this span, but will not propagate up the call tree. The more attributes are
+ * added to the context, the heavier it makes the requests on the wire, so it must be used with
+ * caution.
+ */
+public interface TraceContext {
+
+  /**
+   * Stores Trace Attribute in the context as a key/value pair. Enables powerful distributed context
+   * propagation functionality where arbitrary application data can be carried along the full path
+   * of request execution throughout the system.
+   *
+   * <p>Note 1: attributes are only propagated to the future children of the current trace context.
+   *
+   * <p>Note 2: attributes are sent in-band with every subsequent local and remote calls, so this
+   * feature must be used with care.
+   *
+   * <p>Note 3: keys are case-insensitive, to allow propagation via HTTP headers. Keys must match
+   * regex {@code (?i:[a-z0-9][-a-z0-9]*)}
+   *
+   * @param key attribute key, matching regex {@code (?i:[a-z0-9][-a-z0-9]*)}
+   * @param value value associated with the key.
+   */
+  TraceContext setAttribute(String key, String value);
```

Comment:
I would argue against deletion semantics... it's already a confusing concept since attributes only propagate to *future* children, meaning they're not truly trace-wide (don't affect past children or other descendants of parents ("cousins", etc). Adding deletion into the mix just makes this harder to reason about. Finally, it's easier to add features than subtract them. Thoughts?